Sources/Valet/Internal/Keychain.swift

```diff
@@ -48,216 +48,173 @@ internal final class Keychain {
+            try SecItem.add(attributes: secItemQuery)
         #else
-            
-            if case .success = containsObject(forKey: key, options: options) {
-                return SecItem.update(attributes: [kSecValueData as String: object], forItemsMatching: secItemQuery)
-            } else {
-                secItemQuery[kSecValueData as String] = object
-                return SecItem.add(attributes: secItemQuery)
-            }
+
+        if containsObject(forKey: key, options: options) == errSecSuccess {
+            try SecItem.update(attributes: [kSecValueData as String: object], forItemsMatching: secItemQuery)
+        } else {
+            secItemQuery[kSecValueData as String] = object
+            try SecItem.add(attributes: secItemQuery)
+        }
         #endif
     }
     
     // MARK: Removal
     
-    internal static func removeObject(forKey key: String, options: [String : AnyHashable]) -> SecItem.Result {
+    internal static func removeObject(forKey key: String, options: [String : AnyHashable]) throws {
         guard !key.isEmpty else {
-            ErrorHandler.assertionFailure("Can not set a value with an empty key.")
-            return .error(errSecParam)
+            throw KeychainError.emptyKey
         }
         
         var secItemQuery = options
         secItemQuery[kSecAttrAccount as String] = key
         
-        switch SecItem.deleteItems(matching: secItemQuery) {
-        case .success:
-            return .success
-            
-        case let .error(status):
-            switch status {
-            case errSecInteractionNotAllowed, errSecMissingEntitlement:
-                return .error(status)
-                
-            default:
-                // We succeeded as long as we can confirm that the item is not in the keychain.
-                return .success
-            }
-        }
+        try SecItem.deleteItems(matching: secItemQuery)
     }
     
-    internal static func removeAllObjects(matching options: [String : AnyHashable]) -> SecItem.Result {
-        switch SecItem.deleteItems(matching: options) {
-        case .success:
-            return .success
-            
-        case let .error(status):
-            switch status {
-            case errSecInteractionNotAllowed, errSecMissingEntitlement:
-                return .error(status)
-                
-            default:
-                // We succeeded as long as we can confirm that the item is not in the keychain.
-                return .success
-            }
-        }
+    internal static func removeAllObjects(matching options: [String : AnyHashable]) throws {
+        try SecItem.deleteItems(matching: options)
     }
     
     // MARK: Contains
     
-    internal static func containsObject(forKey key: String, options: [String : AnyHashable]) -> SecItem.Result {
+    internal static func containsObject(forKey key: String, options: [String : AnyHashable]) -> OSStatus {
```

Comment:
Should we be exposing `OSStatus` in the interface here? Seems like `Keychain` abstracts away the underlying error codes everywhere else except here.

Reply:
This is an `internal` method. The call-sites need to inspect the status code in order to make informed decisions. Specifically, the `SecureEnclaveValet` and `SinglePromptSecureEnclaveValet` need to test if an authentication failure error occured, whereas the regular `Valet` does not.

Reply:
Previously, we included the necessary information in `SecItem.Result`. I could theoretically create a new intermediary type (we've deleted `SecItem.Result` in this PR), but since we're not exposing anything publicly that seemed like overkill? I agree that this is a bit of a sharp edge in an otherwise very simple system.

Reply:
Yeah, that's fair. Given it's all internal, it seems alright.